src/jit/morph.cpp

```diff
@@ -1242,158 +1242,96 @@ void fgArgInfo::RemorphReset()
-    unsigned       inx;
-
-    for (inx = 0; inx < argCount; inx++)
-    {
-        curArgTabEntry = argTable[inx];
-        if (curArgTabEntry->argNum == argNum)
-        {
-            break;
-        }
+    bool isLateArg = (curArgTabEntry->lateArgInx != (unsigned)-1);
+    assert((isLateArg && ((node->gtFlags & GTF_LATE_ARG) != 0)) ||
+           (!isLateArg && ((node->gtFlags & GTF_LATE_ARG) == 0)));
 
-        bool     isRegArg;
-        GenTree* argx;
-        if (curArgTabEntry->parent != nullptr)
-        {
-            assert(curArgTabEntry->parent->OperIsList());
-            argx     = curArgTabEntry->parent->Current();
-            isRegArg = (argx->gtFlags & GTF_LATE_ARG) != 0;
-        }
-        else
-        {
-            argx     = curArgTabEntry->node;
-            isRegArg = true;
-        }
-
-        if (isRegArg)
-        {
-            regArgInx++;
-        }
-    }
-    // if this was a nonstandard arg the table is definitive
-    if (curArgTabEntry->isNonStandard)
+    if (curArgTabEntry->parent != nullptr)
     {
-        regNum = curArgTabEntry->regNum;
+        assert(curArgTabEntry->parent->OperIsList());
+        assert(curArgTabEntry->parent->Current() == node);
     }
 
-    assert(curArgTabEntry->argNum == argNum);
-    assert(curArgTabEntry->regNum == regNum);
-    assert(curArgTabEntry->alignment == alignment);
-    assert(curArgTabEntry->parent == parent);
-
     if (curArgTabEntry->node != node)
     {
         GenTree* argx     = nullptr;
         unsigned regIndex = 0;
-
-        /* process the register argument list */
-        for (GenTreeArgList* list = callTree->gtCall.gtCallLateArgs; list; (regIndex++, list = list->Rest()))
+        if (reMorphing)
         {
-            argx = list->Current();
-            assert(!argx->IsArgPlaceHolderNode()); // No place holders nodes are in gtCallLateArgs;
-            if (regIndex == regArgInx)
+            // Find the arg in the late args list.
+            unsigned regArgInx = curArgTabEntry->lateArgInx;
+            for (GenTreeArgList* list = callTree->gtCall.gtCallLateArgs; list; (regIndex++, list = list->Rest()))
             {
-                break;
+                argx = list->Current();
+                assert(!argx->IsArgPlaceHolderNode()); // No place holders nodes are in gtCallLateArgs;
+                if (regIndex == regArgInx)
+                {
+                    break;
+                }
+            }
+            assert(regIndex == regArgInx);
+            if (curArgTabEntry->node != argx)
+            {
+                curArgTabEntry->node = argx;
             }
         }
-        assert(regIndex == regArgInx);
-        assert(regArgInx == curArgTabEntry->lateArgInx);
-
-        if (curArgTabEntry->node != argx)
+        else
         {
-            curArgTabEntry->node = argx;
+            assert(curArgTabEntry->lateArgInx == (unsigned)-1);
+            curArgTabEntry->node = node;
         }
     }
     return curArgTabEntry;
```

Comment:
Why isn't this a `void` return? Seems not useful (and possibly confusing) to always return the input arg.

Reply:
Changed. (Meant to do that before).